src/tests/tcurl_test_tool.c

```diff
@@ -28,188 +28,254 @@
+            return EINVAL;
         }
     }
 
-    DEBUG_CLI_INIT(pc_debug);
-
-    tool_ctx = talloc_zero(NULL, struct tool_ctx);
-    if (tool_ctx == NULL) {
-        DEBUG(SSSDBG_CRIT_FAILURE, "Could not init tool context\n");
-        return 1;
+    if (opt != -1) {
+        poptPrintUsage(pc, stderr, 0);
+        fprintf(stderr, "%s", poptStrerror(opt));
+        return EINVAL;
     }
 
-    inbufs = talloc_zero_array(tool_ctx, struct sss_iobuf *, MAXREQ);
-    if (inbufs == NULL) {
-        talloc_zfree(tool_ctx);
-        return 1;
+    return EOK;
+}
+
+static errno_t
+prepare_requests(TALLOC_CTX *mem_ctx,
+                 poptContext pc,
+                 struct tool_options *opts,
+                 struct tcurl_request ***_requests,
+                 size_t *_num_requests)
+{
+    struct tcurl_request **requests;
+    struct sss_iobuf *body;
+    const char **headers;
+    const char *arg;
+    const char *url;
+    errno_t ret;
+    size_t i;
+
+    static const char *default_headers[] = {
+        "Content-type: application/octet-stream",
+        NULL,
+    };
+
+    requests = talloc_zero_array(mem_ctx, struct tcurl_request *, MAXREQ + 1);
+    if (requests == NULL) {
+        return ENOMEM;
     }
 
-    while ((extra_arg_ptr = poptGetArg(pc)) != NULL) {
-        switch (req_type) {
+    headers = opts->headers == NULL ? default_headers : opts->headers;
+
+    i = 0;
+    while ((arg = poptGetArg(pc)) != NULL) {
+        if (i >= MAXREQ) {
+            fprintf(stderr, _("Too many requests!\n"));
+            ret = EINVAL;
+            goto done;
+        }
+
+        switch(opts->method) {
         case TCURL_HTTP_GET:
         case TCURL_HTTP_DELETE:
-            urls[n_reqs++] = extra_arg_ptr;
+            url = arg;
+            body = NULL;
             break;
         case TCURL_HTTP_PUT:
-            if (urls[n_reqs] == NULL) {
-                urls[n_reqs] = extra_arg_ptr;
-            } else {
-                inbufs[n_reqs] = sss_iobuf_init_readonly(
-                                              inbufs,
-                                              (uint8_t *) discard_const(extra_arg_ptr),
-                                              strlen(extra_arg_ptr));
-                if (inbufs[n_reqs] == NULL) {
-                    DEBUG(SSSDBG_CRIT_FAILURE, "Could not init input buffer\n");
-                    talloc_zfree(tool_ctx);
-                    return 1;
-                }
-                n_reqs++;
+        case TCURL_HTTP_POST:
+            url = arg;
+
+            arg = poptGetArg(pc);
+            if (arg == NULL) {
+                body = NULL;
+                break;
+            }
+
+            body = sss_iobuf_init_readonly(requests,
+                                           discard_const_p(uint8_t, arg),
+                                           strlen(arg));
+            if (body == NULL) {
+                ret = ENOMEM;
+                goto done;
             }
             break;
```

Comment:
I assume that gcc does not like missing default and there is a Wmaybe-unitialized warning
```
../sssd/src/tests/tcurl_test_tool.c: In function ‘main’:
../sssd/src/tests/tcurl_test_tool.c:181:23: error: ‘url’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
         requests[i] = tcurl_http(requests, opts->method, opts->socket_path,
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                  url, headers, body);
                                  ~~~~~~~~~~~~~~~~~~~
../sssd/src/tests/tcurl_test_tool.c:131:17: note: ‘url’ was declared here
     const char *url;
                 ^~~
../sssd/src/tests/tcurl_test_tool.c:181:23: error: ‘body’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
         requests[i] = tcurl_http(requests, opts->method, opts->socket_path,
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                  url, headers, body);
                                  ~~~~~~~~~~~~~~~~~~~
../sssd/src/tests/tcurl_test_tool.c:128:23: note: ‘body’ was declared here
     struct sss_iobuf *body;
                       ^~~~

```